### DISCUSSION TITLE
ACAP program installation process

### DISCUSSION THREAD

**Discussion by sinnrrr** [AxisCommunications/discussions#102] [ACAP] [answered]
I noticed there is a difference between how apps are being installed on the camera.

I had a situation, where logs from one app, haven't been captured, while from another everything worked. I decided to view to `systemd` unit configuration for these apps and found out that one of them (where logs didn't capture) was launched using `init.d` and another one normally (full path to binary). *That means that everything from `stdout` goes to `init.d` and **never reaches** `systemd`, which is why we cannot see logs, as **they are not put into the journal**.*

I have noticed one difference between these apps: the `manifest.json` file is missing in the app package, the logs of which does not being captured. Does this mean that the installer acts differently based on that? In which cases it includes `init.d` in the unit configuration file of the app, and why use `init.d` at all if you're using `systemd`?

**Comment by mattias-kindborg-at-work** [comment]
Hi @sinnrrr. Are these apps publicly accessible, can I inspect them?

`manifest.json` is a continuation of `package.conf`, for more information see the [ACAP version 3 documentation](https://help.axis.com/acap-3-developer-guide#manifest-file). ACAP applications built with newer versions of the ACAP version 3 SDK or ACAP version 4 SDK will use `manifest.json` as a description of the package. Logs are accessible via `http://<ip>/axis-cgi/admin/systemlog.cgi?appname=<app name>` where `<ip>` is the IP address of the Axis device and `<app name>` is the name/id of the application. Could you please verify that logs for both applications are available using this URI?

**Reply by sinnrrr** [reply]
@mattias-kindborg-at-work  Hi!
No, they are not publicly accessible, unfortunately :(
Yes, I can access this URL, but there are no logs:
```
----- Contents of SYSTEM_LOG for 'rr_axis_streamer' -----
```

I've detected that `systemd` service of this app uses `init.d` to launch the app:
```
[Unit]
Description=rr_axis_streamer
After=acap-pre.target
Requires=acap-pre.target
After=respawnd.service
Requires=respawnd.service
[Service]
Slice=extension-acap.slice
Type=oneshot
RemainAfterExit=yes
ExecStart=/etc/init.d/sdkrr_axis_streamer start
ExecStop=/etc/init.d/sdkrr_axis_streamer stop
[Install]
WantedBy=acap.target
```
I think this is the reason why logs are not captured into `journalctl`.

**Reply by sinnrrr** [reply]
I have another camera that have the same app installed there, the logs **ARE** available there and the `systemd` service looks as follows:
```
[Unit]
Description=RR Axis Streamer
After=acap-pre.target
Requires=acap-pre.target
[Service]
Slice=extension-acap.slice
RemainAfterExit=yes
ExecStart=/usr/local/packages/rr_axis_streamer/rr_axis_streamer
Restart=always
User=root
Group=root
[Install]
WantedBy=acap.target
```

**Reply by mattias-kindborg-at-work** [reply]
Given that this is the same installed ACAP application, does the firmware version of the two cameras differ?

**Reply by sinnrrr** [reply]
Yeah, it seems like one camera had `10.12.xx`, whilst the other had `10.4.xx` (this one didn't show the logs).
Upgrading the firmware worked!

**Reply by daniel-falk** [reply, marked as answer]
I think the standard for cameras is still to use `init.d` (even in FW 11.x) which means that `stdout` is not captured by the system log? To make debugging easier we use a library that logs both to `stdout` and to the system log, like this:
```c
/* Log a message both to the system logger and to stdout */
void fapp_logger_log(int priority, const char *fmt, ...) {
  va_list args;

  /* Log to stdout */
  va_start(args, fmt);
  printf("%s: ", fapp_logger_level_to_str(priority));
  vprintf(fmt, args);
  printf("\n");
  va_end(args);

  /* Log to system logger */
  va_start(args, fmt);
  vsyslog(priority, fmt, args);
  va_end(args);
}
```

Some third party libraries logs only to `stdout` in which case it is much harder to capture the logs.. Manually redirecting the `stdout` file descriptor is of course possible...

**Reply by dsoltysiuk** [reply]
Thanks! This makes sense.
